exp/internal/webhooks/ipaddress.go

```diff
@@ -0,0 +1,169 @@
+		Complete()
+}
+
+// +kubebuilder:webhook:verbs=create;update;delete,path=/validate-cluster-x-k8s-io-v1beta1-ipaddress,mutating=false,failurePolicy=fail,matchPolicy=Equivalent,groups=cluster.x-k8s.io,resources=ipaddresses,versions=v1beta1,name=validation.ipaddress.cluster.x-k8s.io,sideEffects=None,admissionReviewVersions=v1;v1beta1
+// +kubebuilder:rbac:groups=cluster.x-k8s.io,resources=ipaddressclaims,verbs=get;list;watch
+
+// IPAddress implements a validating webhook for IPAddress.
+type IPAddress struct {
+	Client client.Reader
+}
+
+var _ webhook.CustomValidator = &IPAddress{}
+
+// ValidateCreate implements webhook.CustomValidator.
+func (webhook *IPAddress) ValidateCreate(ctx context.Context, obj runtime.Object) error {
+	ip, ok := obj.(*expv1.IPAddress)
+	if !ok {
+		return apierrors.NewBadRequest(fmt.Sprintf("expected an IPAddress but got a %T", obj))
+	}
+	return webhook.validate(ctx, ip)
+}
+
+// ValidateUpdate implements webhook.CustomValidator.
+func (webhook *IPAddress) ValidateUpdate(ctx context.Context, oldObj, newObj runtime.Object) error {
+	oldIP, ok := oldObj.(*expv1.IPAddress)
+	if !ok {
+		return apierrors.NewBadRequest(fmt.Sprintf("expected an IPAddress but got a %T", oldObj))
+	}
+	newIP, ok := newObj.(*expv1.IPAddress)
+	if !ok {
+		return apierrors.NewBadRequest(fmt.Sprintf("expected an IPAddress but got a %T", newObj))
+	}
+
+	if !reflect.DeepEqual(oldIP.Spec, newIP.Spec) {
+		return field.Forbidden(field.NewPath("spec"), "the spec of IPAddress is immutable")
+	}
+	return nil
+}
+
+// ValidateDelete implements webhook.CustomValidator.
+func (webhook *IPAddress) ValidateDelete(ctx context.Context, obj runtime.Object) error {
+	return nil
+}
+
+func (webhook *IPAddress) validate(ctx context.Context, ip *expv1.IPAddress) error {
+	allErrs := field.ErrorList{}
+
+	addr, err := netaddr.ParseIP(ip.Spec.Address)
+	if err != nil {
+		allErrs = append(allErrs,
+			field.Invalid(
+				field.NewPath("spec.address"),
+				ip.Spec.Address,
+				"not a valid IP address",
+			))
+	}
+
+	if ip.Spec.Prefix < 0 {
+		allErrs = append(allErrs,
+			field.Invalid(
+				field.NewPath("spec.prefix"),
+				ip.Spec.Prefix,
+				"prefix cannot be negative",
+			))
+	}
+	if addr.Is4() && ip.Spec.Prefix > 32 {
+		allErrs = append(allErrs,
+			field.Invalid(
+				field.NewPath("spec.prefix"),
+				ip.Spec.Prefix,
+				"prefix is too large for an IPv4 address",
+			))
+	}
+	if addr.Is6() && ip.Spec.Prefix > 128 {
+		allErrs = append(allErrs,
+			field.Invalid(
+				field.NewPath("spec.prefix"),
+				ip.Spec.Prefix,
+				"prefix is too large for an IPv6 address",
+			))
+	}
+
+	_, err = netaddr.ParseIP(ip.Spec.Gateway)
+	if err != nil {
+		allErrs = append(allErrs,
+			field.Invalid(
+				field.NewPath("spec.gateway"),
+				ip.Spec.Gateway,
+				"not a valid IP address",
+			))
+	}
+
+	claim := &expv1.IPAddressClaim{}
+	err = webhook.Client.Get(ctx, types.NamespacedName{Name: ip.Spec.ClaimRef.Name, Namespace: ip.ObjectMeta.Namespace}, claim)
```

Comment:
This creates a dependency between CR, defining the order they should be created; this order should be taken into consideration in operations like move or valero backup restore.
Given that this is complicated, and not really git-ops friendly, is it possible to remove the check based on claim ref from validation? what are the risk that we are introducing for the users?

Reply:
Valid point.
But I don't think GitOps has to be taken into consideration. IPAddresses are supposed to be created by an IPAM provider. The exist to fulfil a claim, and the provider would only create them after the claim is created.
Backups are another topic though.

We can soften it up to only verify if the pool matches on both resources when the referenced Claim is found. But that way you could, in theory, create an IPAddress first, and then a Claim that matches the reference, but references a different Pool.
Since these resources are created by other providers, at least that's the main goal, the validation only ensures providers are behaving correctly. Users probably do not interact with it directly, so we could argue that we don't need as much validation here.

Reply:
Some background info: We had a similar issue with Cluster / ClusterClass validation. We ended up with keeping the validation / ordering constraint but hard-coded the restore order in Velero.

Reply:
I think that would be a good approach here as well. It shouldn't be relevant apart from backups imo.

Reply:
The alternative would be to just document that.
Is this a decision then to keep the validation in place? Then I'd propose to PR it to Velero after this is merged.

I mean, cluster-api is an official k8s project. It's fine to have built-in support for those in tools like Velero. Would be interesting to know where they draw the line. Maybe there is none, doesn't hurt to have more entries in that list I guess if it makes using Velero easier.

Reply:
To be honest. I'm personally fine with both. So I leave it to you and Fabrizio to decide.

I think we shouldn't make CAPI functionality dependent on external tools. Who knows which other GitOps tools or backup tools might have an issue with that or not. (we probably unfortunately also have to take a look at clusterctl backup/restore)

Velero also allows specifying the order via command line option, which could be a workaround for Velero until the IP types have been added to the hard-coded order.

As I said I'm fine with both options, but one straightforward compromise could be to only validate the ref if it exists in this PR and then create an issue to follow-up on: Velero, clusterctl backup/restore and then introduce the "hard" constraint in the webhook again.

Reply:
+1 to remove the check now if this does not introduce problems.
valero is an external tool, but it is the solution we recommend for backup restore, so it is in our interest to ensure compatibility.
But valero is only half of the problem; in order to get move work, then we need to ensure that owner ref are properly set, and this is slightly more complicated given that the IPAM logic will be distributed between providers.

Reply:
I'll resolve this one given we have https://github.com/kubernetes-sigs/cluster-api/pull/6313#discussion_r914977795